dask/bag/core.py

```diff
@@ -171,21 +189,21 @@ def to_textfiles(b, path, name_function=None, compression='infer',
 
     >>> b_dict.map(json.dumps).to_textfiles("/path/to/data/*.json")  # doctest: +SKIP
     """
-    from dask import delayed
-    (writes,names) = write_bytes(b.to_delayed(), path, name_function, compression,
-                                 encoding=encoding, **(storage_options or {}))
+    mode = 'wb' if encoding is None else 'wt'
+    files = open_files(path, compression=compression, mode=mode,
+                       encoding=encoding, name_function=name_function,
+                       num=b.npartitions, **(storage_options or {}))
 
-    # Use Bag optimizations on these delayed objects
-    dsk = ensure_dict(delayed(writes).dask)
-    dsk2 = Bag.__dask_optimize__(dsk, [w.key for w in writes])
-    out = [Delayed(w.key, dsk2) for w in writes]
+    name = 'to-textfiles-' + uuid.uuid4().hex
+    dsk = {(name, i): (_to_textfiles_chunk, (b.name, i), f)
+           for i, f in enumerate(files)}
+    out = type(b)(merge(dsk, b.dask), name, b.npartitions)
```

Comment:
This makes a bag?
So if compute is false, we unwrap the bag and give back the delayed `_to_textfiles_chunk`s?

Reply:
Correct. But in that case we only call `__dask_optimize__` once, rather than twice as we're currently doing. This potentially allows more inlining/fusing to happen, and avoids doing extra work. Seemed cleaner overall to me.